docs/0005-hashed-n-tuple-id-layout.md

```diff
@@ -0,0 +1,235 @@
+# OCFL Community Extension 0005: Hashed Truncated N-tuple Trees with Encapsulating Directory for OCFL Storage Hierarchies
+
+  * **Extension ID:** 0005-hashed-n-tuple-id-layout
+  * **Authors:** Ben Cail
+  * **Minimum OCFL Version:** 1.0
+  * **Obsoletes:** n/a
+  * **Obsoleted by:** n/a
+
+## Overview
+
+This storage root extension describes how to safely map OCFL object identifiers containing any characters to OCFL object root directories.
+
+Using this extension, OCFL object identifiers are hashed and encoded as hex strings (all letters lower-case). These digests are then divided into _N_ n-tuple segments, which are used to create nested paths under the OCFL storage root. Finally, the OCFL object identifier is URL-encoded to create a directory name for the OCFL object root.
+
+The n-tuple segments approach allows OCFL object identifiers to be evenly distributed across the storage hierarchy. The maximum number of files under any given directory is controlled by the number of characters in each n-tuple, and the tree depth is controlled by the number of n-tuple segments each digest is divided into. The encoded encapsulation directory name provides visibility into the object identifier from the file path, but it does require that the encoded object identifier not be longer than 255 characters (because of filesystem limitations on the length of a directory name).
```

Comment:
I'm thinking about changing the behaviour to be able handle long object IDs (instead of throwing an exception).
option 1: could just switch to using the digest as the encapsulation directory if the encoded object id is too long
option 2: could truncate the encoded object id and add the digest at the end, so that the total length is 255 chars (eg. something like longobjectidtruncaff75534492485eabb39f86356728884e).

Thoughts? Option 1 would be simpler, but would lose all object id context. Option 2 is a bit more complicated, but preserved a chunk of the object id.

Reply:
@bcail, I don't have a great answer for this.

Option 1 is odd because, if you're browsing the filesystem, you don't definitively know if the encapsulation directory is a hashed or unhashed value.  Additionally, it has a collision problem if you have an unhashed object id that equals the hashed value of a different id.

Option 2 could work if you did something like, if encoded object ids are longer than 100 characters, then the unencoded object id is hashed and used to construct an encapsulation directory in the format of "<encoded-object-id[0:100]>-<object-id-digest>". There is got-ya here, you can't just truncate the encoded object id to N characters because it might split a character encoding. But, perhaps you don't care if the string can be unencoded because it's already truncated? It does feel a little jarring, but I also don't have a better suggestion right now.

I'm wondering how necessary behavior is. I would assume that most users would know the nature of their IDs when they setup a repository and would know whether or not this extension would work for them. I would also assume that, in most cases, people won't be using ids that would be encoded strings longer than 255 characters. Fedora 6 is an example of where this assumption cannot be made, and, as such, it is not going to use this storage layout.

Reply:
Those are good points about the option 1: a 128-character directory name could be a hash of a long object id, or an unhashed object id, and whichever it is, it could conflict with the other type of object.

For option 2, I'm not sure I'm too concerned about splitting up a character encoding, because you already can't get back to the full object id (as you mentioned).

I think for now I'll pursue option 2.